### PR TITLE
wip: Twitter OAuth integration

### LINKED ISSUE #5363: Implement additional signups
Please implement additional signups as follows for speakers and attendees:

* [ ] If you would like your information to be shared with potential recruiters, please check this box. By checking, you allow the Linux Foundation to share your information with potential recruiters. 

* [ ] By checking here, you opt-in to being listed in the event's online attendee directory. Some of your registration data will be made available to other event attendees in the directory (name, title, and company name only). 

* [ ] By checking here, you opt-in to receiving communications from the event's sponsors regarding their own news and messages (sent by The Linux Foundation; your contact information will not be shared).

* [ ] By checking here, you opt-in to receiving occasional communications from The Linux Foundation relating to our other open source projects, events, and services like training and certifications. If so, you can unsubscribe from email communications at any time using the email's Unsubscribe link. Please note that The Linux Foundation may contact you regarding events and other services you actually register for, even if you do not check this box or if you use a marketing email's Unsubscribe link.

### PR DESCRIPTION
**Pull request by mrsaicharan1**
Fixes #5363 

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [ ] All the functions created/modified in this PR contain relevant docstrings.

#### Changes proposed in this pull request:

- Added Twitter-OAuth login functionality using Flask-Dance
- Additional checks for account info and authorization
- Schema changes for Twitter account integration

### CODE REVIEW

**Comment by mrsaicharan1** [comment]
@iamareebjamal The problem isn’t migration heads because I created a new PR, new branch and prior yo everyting, committed only the database changes and migration file(where it passed Travis build). When I went on to committing the auth.py changes, it throws an error saying that the settings relation doesn’t exist. Is there a workaround for this?

File: requirements/common.txt [hunk 1]
```diff
@@ -4,7 +4,7 @@ Flask-Script>=2.0.5,<2.1
 Flask-SQLAlchemy>=2.1,<2.2
 Flask-Migrate>=2.0.0,<3
 Flask-Login>=0.4,<1
-Flask_Dance>=1.3.0
+Flask-Dance>=1.3.0
```
Comment (iamareebjamal) [inline requirements/common.txt]: See the format in which other dependencies are pinned

Comment (mrsaicharan1) [reply requirements/common.txt]: Will resolve this. Just checking for the main reason for Travis fail.

File: app/api/auth.py [hunk 2]
```diff
@@ -23,6 +24,10 @@
 from app.models.user import User
 
 auth_routes = Blueprint('auth', __name__, url_prefix='/v1/auth')
+api_key = TwitterOAuth.get_client_id()
```
Comment (CosmicCoder96) [inline app/api/auth.py]: @iamareebjamal these calls which require fetching values from the settings table are being placed before the app is properly initialized I think. If dummy values are passed to the 
`make_twitter_blueprint(api_key=dummmy_api_key, api_secret=dummy_api_secret)`
the build continues, can you confirm this? Also, I think this is not the correct file to register this blueprint, rather should be registered with other blueprints.

**Comment by mrsaicharan1** [comment]
Yes, @CosmicCoder96 will change the blueprint registration to the
associated __init__.py file.

File: app/api/auth.py (hunk 2 above)
Comment (mrsaicharan1) [reply app/api/auth.py]: @CosmicCoder96 How are we going to put in the api key and secret for now?

**Comment by mrsaicharan1** [comment]
Blueprint registration done in `__init__.py`

**Comment by CosmicCoder96** [comment]
> Yes, @CosmicCoder96 will change the blueprint registration to the associated __init__.py file. That's file where we're registering routes right?

Yes, but I am not sure if it will fix the reason for failed build. The call to `get_settings()` before initialisation of app is the problem.

**Comment by mrsaicharan1** [comment]
> > Yes, @CosmicCoder96 will change the blueprint registration to the associated **init**.py file. That's file where we're registering routes right?
> 
> Yes, but I am not sure if it will fix the reason for failed build. The call to `get_settings()` before initialisation of app is the problem.

How is it happening before initialization. Doesn't `init.py` initialize the app?

**Comment by mrsaicharan1** [comment]
@CosmicCoder96 @iamareebjamal What if we store/obtain the keys from an external source such as Amazon S3?

**Comment by iamareebjamal** [comment]
1. How would storing it in S3 help?
2. S3 is a file storage 

Any top-level declaration in any python file is run on any import, not just a single __init__.py

In fact, all imported files top-level declarations will run first, and the app's __init__.py will run last

**Comment by mrsaicharan1** [comment]
Storing the keys in a txt file on S3. But I figured that it would require a
lot of time and load the network requests. Nevermind. I  initialized the
database with a manual request before getting the keys from Setting.

File: app/api/auth.py [hunk 3]
```diff
@@ -251,9 +274,7 @@ def change_password():
         return NotFoundError({'source': ''}, 'User Not Found').respond()
     else:
         if user.is_correct_password(old_password):
-            if len(new_password) < 8:
```
Comment (iamareebjamal) [inline app/api/auth.py]: Why removed?

File: app/api/auth.py [hunk 4]
```diff
@@ -23,6 +24,11 @@
 from app.models.user import User
 
 auth_routes = Blueprint('auth', __name__, url_prefix='/v1/auth')
+with app.app_context():
+     db.init_app(app)
```
Comment (iamareebjamal) [inline app/api/auth.py]: I'm not sure this is a good idea

File: app/api/auth.py (hunk 3 above)
Comment (mrsaicharan1) [reply app/api/auth.py]: Just saw this. Minor typo. Fixing it.

File: app/api/auth.py (hunk 4 above)
Comment (mrsaicharan1) [reply app/api/auth.py]: I guess until initializing the app under a manual context is ok(The application context is created and destroyed as necessary.), as it's working out of the application context.

Comment (mrsaicharan1) [reply app/api/auth.py]: It only initializes the database under a manual context. Um, is this wrong?

Comment (iamareebjamal) [reply app/api/auth.py]: It should not be needed

File: app/api/auth.py [hunk 5]
```diff
@@ -23,6 +24,11 @@
 from app.models.user import User
 
 auth_routes = Blueprint('auth', __name__, url_prefix='/v1/auth')
+with app.app_context():
+    db.init_app(app)
+    db.create_all()
+    twitter_blueprint = make_twitter_blueprint(api_key=TwitterOAuth.get_client_id(),
```
Comment (iamareebjamal) [inline app/api/auth.py]: Remove this

Comment (mrsaicharan1) [reply app/api/auth.py]: Will remove it. But Travis fails upon doing that. How do we make it pass?

Comment (iamareebjamal) [reply app/api/auth.py]: Initialize this in create_app

Comment (iamareebjamal) [reply app/api/auth.py]: Or see any sample app of Flask Dance

**Comment by mrsaicharan1** [comment]
Will do it.

**Comment by mrsaicharan1** [comment]
I've actually tried doing it in creare_app() as well, but still, build
fails.

**Comment by mrsaicharan1** [comment]
`AssertionError: A name collision occurred between blueprints <flask_dance.consumer.oauth1.OAuth1ConsumerBlueprint object at 0x7fe014b0d668> and <flask_dance.consumer.oauth1.OAuth1ConsumerBlueprint object at 0x7fe014dfb748>. Both share the same name "twitter". Blueprints that are created on the fly need unique names.
`
@iamareebjamal

**Comment by iamareebjamal** [comment]
This is because dev server is started twice. See some sample code of the library about where it is initialized

**Comment by mrsaicharan1** [comment]
@iamareebjamal Yes, will checking out the example implementations.

**Comment by mariobehling** [comment]
Priority low: Delay

**Pull request closed without merging**